### DISCUSSION TITLE
Material property and initial Aux variable values from an input (ASCII) file

### DISCUSSION THREAD

**Discussion by chakra34** [idaholab/moose#24679] [Q&A General] [answered]
Hi Moose community,

I wanted to ask if there is a way to provide the value of material properties as an input file. For example I have a cylindrical geometry where I want to provide the temperature field, Young's modulus and Poisson's ratio from an input file that has the mesh coordinates and the corresponding values of these material properties and temperature (treated as an aux variable for this purpose).

Is there already something in MOOSE which allows to do this?

Thanks,
Best,

Aritra

**Comment by GiudGiud** [comment]
Hello

So those properties are solely a function of space?
If so:
This Function
https://mooseframework.inl.gov/source/functions/PiecewiseConstantFromCSV.html
then this material
https://mooseframework.inl.gov/source/materials/GenericFunctionMaterial.html

Guillaume

**Reply by chakra34** [reply]
Thanks for the reply! I guess I am a bit confused on how we are incorporating this information. As an example I had a csv file with x,y,z coordinates and the temperature and I wanted to use this as the function value for the temperature aux variable.

So first few rows of my csv file looks like:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/CHAKRA~1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/CHAKRA~1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



5.00E-02 | 5.22E-02 | 5.22E-02 | 10.439
-- | -- | -- | --
0.1 | 5.22E-02 | 5.22E-02 | 10.439
0.15 | 5.22E-02 | 5.22E-02 | 10.439
0.2 | 5.22E-02 | 5.22E-02 | 10.439
0.25 | 5.22E-02 | 5.22E-02 | 10.439
0.3 | 5.22E-02 | 5.22E-02 | 10.439
0.35 | 5.22E-02 | 5.22E-02 | 10.439
0.4 | 5.22E-02 | 5.22E-02 | 10.439
0.45 | 5.22E-02 | 5.22E-02 | 10.439
0.5 | 5.22E-02 | 5.22E-02 | 10.439
0.55 | 5.22E-02 | 5.22E-02 | 10.439
0.6 | 5.22E-02 | 5.22E-02 | 10.439
0.65 | 5.22E-02 | 5.22E-02 | 10.439
0.7 | 5.22E-02 | 5.22E-02 | 10.439
0.75 | 5.22E-02 | 5.22E-02 | 10.439
0.8 | 5.22E-02 | 5.22E-02 | 10.439
0.85 | 5.22E-02 | 5.22E-02 | 10.439
0.9 | 5.22E-02 | 5.22E-02 | 10.439
0.95 | 5.22E-02 | 5.22E-02 | 10.439
5.00E-02 | 5.21E-02 | 0.10439 | 20.878
0.1 | 5.21E-02 | 0.10439 | 20.878



</body>

</html>


the first 3 are the X,Y,Z coordinates and the last column is the temperature of a box geometry with coordinates ranging from [0,1]m in all X,Y, and Z direction. 
So to do this I used the "nearest" and used the following:

[UserObjects]
  [reader_nearest]
    type = PropertyReadFile
    prop_file_name = 'temperature_ansys_modified.csv'
    read_type = 'voronoi'
    nprop = 1
    nvoronoi = 1
  []
[]

[Functions]
  [temp_nearest]
    type = PiecewiseConstantFromCSV
    read_prop_user_object = 'reader_nearest'
    read_type = 'voronoi'
    column_number = '3'
  []
  
[]

And added in the temperature aux variable:

[AuxKernels]
  [set_temp]
     variable = temp
     type = FunctionAux
     function = temp_nearest #temp_function
  []
[]

But I don't see the temperature getting input.

Meanwhile, I was also looking at the PropertyReadFile.C but and saw some "voronoi" options which I am not sure what to do when I don't have a polygonized grain structure but rather a simple block geometry.

**Reply by GiudGiud** [reply]
this is wrong
```
nvoronoi = 1
```
you have plenty of nearest point regions, you need to have the number of CSV rows as the nvoronoi

**Reply by chakra34** [reply]
Tried but it did not work. I am attaching the full input file:
```
[Mesh]
  type = GeneratedMesh
  dim = 3
  nx = 20
  ny = 20
  nz = 20
  xmin = 0.0
  xmax = 1.0
  ymin = 0.0
  ymax = 1.0
  zmin = 0.0
  zmax = 1.0
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[UserObjects]
  [reader_nearest]
    type = PropertyReadFile
    prop_file_name = 'temperature_ansys_modified.csv'
    read_type = 'voronoi'
    nprop = 1
    nvoronoi = 9261
  []
[]

[Functions]
  [./topfunc]
    type = ParsedFunction
    value = '1e-4*t' # ramped displacement BC
  [../]

  [temp_nearest]
    type = PiecewiseConstantFromCSV
    read_prop_user_object = 'reader_nearest'
    read_type = 'voronoi'
    column_number = '3'
  []
  
[]

[BCs]

  [./bot_fix_y]
    type = DirichletBC
    variable = disp_y
    boundary = 'bottom'
    value = 0
  [../]

  [./top_disp_y]
    type = FunctionDirichletBC
    variable = disp_y
    boundary = 'top'
    function = topfunc
  [../]

  [./left_fix_x]
    type = DirichletBC
    variable = disp_x
    boundary = 'left'
    value = 0
  [../]

  [./front_fix_z]
    type = DirichletBC
    variable = disp_z
    boundary = 'front'
    value = 0
  [../]

[]

[Modules/TensorMechanics/Master/All]
  strain = FINITE
  incremental = true
  add_variables = true
  temperature = temp
  eigenstrain_names = eigenstrain  
  generate_output = 'stress_xx stress_yy stress_zz stress_xy stress_yz stress_zx'
  use_automatic_differentiation = false
[]
[Materials]
  [elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 150e9
    poissons_ratio = 0.3
  []

  [./stress]
    type = ComputeFiniteStrainElasticStress
  [../]

  [./thermal_expansion_strain]
    type = ComputeThermalExpansionEigenstrain
    thermal_expansion_coeff = 1.0e-4
    stress_free_temperature = 100.0
    temperature = temp
    eigenstrain_name = eigenstrain
  [../]

[]


[AuxVariables]
  [temp]
    order  = FIRST
    family = LAGRANGE
  []

[]

[AuxKernels]
  [set_temp]
     variable = temp
     type = FunctionAux
     function = temp_nearest 
  []
[]


[Preconditioning]
  [smp]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Transient

  solve_type = 'Newton'

  start_time = 0.0
  dt    = 0.1
  dtmax = 1.0
  dtmin = 1e-31
  nl_abs_tol = 1e-6
  end_time = 0.1
 []

[Outputs]
  print_linear_residuals = true
  perf_graph = true
  csv = true
  [./out]
    type = Exodus
    elemental_as_nodal = true
  [../]
  [./console]
    type = Console
    max_rows = 100
  [../]
[]
```
And my csv file has 9261 rows which are the total number of nodes in my block.

**Reply by GiudGiud** [reply]
what does it look like now?

**Reply by chakra34** [reply]
The temperature is still zero

![image](https://github.com/idaholab/moose/assets/22983992/20b64748-0d2d-4c3f-917d-afefd78a7ea9)

**Reply by GiudGiud** [reply]
```  [set_temp]
     variable = temp
     type = FunctionAux
     function = temp_nearest 
     execute_on = INITIAL
  []
```

or even better, use a FunctionIC

**Reply by chakra34** [reply]
the execute_on = Initial did not work (still getting zero temp); will try FunctionIC next

**Reply by chakra34** [reply]
Hi @GiudGiud I tried IC, but it seems it is not even reading anything from the input file, and I am still getting zero temperatures. I am attaching the input file and the csv, could you please see if there is something wrong with the csv format ?
[temperature_ansys_modified.csv](https://github.com/idaholab/moose/files/11757739/temperature_ansys_modified.csv)


The corresponding MOOSE input:
```
[Mesh]
  type = GeneratedMesh
  dim = 3
  nx = 20
  ny = 20
  nz = 20
  xmin = 0.0
  xmax = 1.0
  ymin = 0.0
  ymax = 1.0
  zmin = 0.0
  zmax = 1.0
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[UserObjects]
  [reader_nearest]
    type = PropertyReadFile
    prop_file_name = 'temperature_ansys_modified.csv'
    read_type = 'voronoi'
    nprop = 1
    nvoronoi = 9261
  []
[]

[Functions]
  [./topfunc]
    type = ParsedFunction
    value = '1e-4*t' # ramped displacement BC
  [../]

  [temp_nearest]
    type = PiecewiseConstantFromCSV
    read_prop_user_object = 'reader_nearest'
    read_type = 'voronoi'
    column_number = '3'
  []
  
[]

[BCs]

  [./bot_fix_y]
    type = DirichletBC
    variable = disp_y
    boundary = 'bottom'
    value = 0
  [../]

  [./top_disp_y]
    type = FunctionDirichletBC
    variable = disp_y
    boundary = 'top'
    function = topfunc
  [../]

  [./left_fix_x]
    type = DirichletBC
    variable = disp_x
    boundary = 'left'
    value = 0
  [../]

  [./front_fix_z]
    type = DirichletBC
    variable = disp_z
    boundary = 'front'
    value = 0
  [../]

[]

[ICs]
  [nearest]
    type = FunctionIC
    variable = 'temp'
    function = 'temp_nearest'
  []

[]


[Modules/TensorMechanics/Master/All]
  strain = FINITE
  incremental = true
  add_variables = true
  temperature = temp
  eigenstrain_names = eigenstrain  
  generate_output = 'stress_xx stress_yy stress_zz stress_xy stress_yz stress_zx'
  use_automatic_differentiation = false
[]
[Materials]
  [elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 150e9
    poissons_ratio = 0.3
  []

  [./stress]
    type = ComputeFiniteStrainElasticStress
  [../]

  [./thermal_expansion_strain]
    type = ComputeThermalExpansionEigenstrain
    thermal_expansion_coeff = 1.0e-4
    stress_free_temperature = 100.0
    temperature = temp
    eigenstrain_name = eigenstrain
  [../]

[]


[AuxVariables]
  [temp]
    order  = FIRST
    family = LAGRANGE
  []

[]


[Preconditioning]
  [smp]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Transient

  solve_type = 'Newton'

  start_time = 0.0
  dt    = 0.1
  dtmax = 1.0
  dtmin = 1e-31
  nl_abs_tol = 1e-6
  end_time = 0.1
 []

[Outputs]
  print_linear_residuals = true
  perf_graph = true
  csv = true
  [./out]
    type = Exodus
    elemental_as_nodal = true
  [../]
  [./console]
    type = Console
    max_rows = 100
  [../]
[]

```
Does something seem off ?

**Reply by GiudGiud** [reply]
Hello

Sorry for the delay.
Something is off indeed, only the first column is being read.
I ll try to fix this

Guillaume

**Reply by chakra34** [reply]
No worries at all @GiudGiud ! Thank you so much for looking into it!! Really appreciate it!

-Aritra

**Reply by GiudGiud** [reply]
Hello

I m digging through it. The CSV reader utility is skipping the first column for some reason, and using the data as the Z coordinate instead.

While I work on the fix on that core utility, if you just want to get started, you should be able to add a "fake" column at the beginning of every row. Then x,y,z then your data

My guess rn is that it s expecting row labels. It needs to be smarter than this. 

Guillaume

**Reply by GiudGiud** [reply, marked as answer]
You have windows line endings on your file.

This breaks the csv reader which is not able to parse the entire lines. 

You can use:
- dos2unix temperature_ansys_modified.csv  (on linux)
EDIT: I m still digging into this and maybe only the latter option works
or
- sed -e 's/\r//g' temperature_ansys_modified.csv &> temperature_ansys_modified_clean.csv

to remove them

**Reply by chakra34** [reply]
Hi @GiudGiud thanks a lot! Yeah adding blank columns in-front of each row indeed works to get the right temperature profiles. So the current csv format looks like:

...
5.00E-02,,5.22E-02,,5.22E-02,,10.439
0.1,,5.22E-02,,5.22E-02,,10.439
0.15,,5.22E-02,,5.22E-02,,10.439
0.2,,5.22E-02,,5.22E-02,,10.439
...

two delimiters before each of the column.
